src/expertsystem/reaction/topology.py

```diff
@@ -315,19 +309,28 @@ class StateTransitionGraph(Topology, Generic[_EdgeType]):
 
     def __init__(
         self,
-        nodes: Optional[Set[int]] = None,
-        edges: Optional[Dict[int, Edge]] = None,
+        topology: Topology,
```

Comment:
I wonder if it's worth using `attr.s` here, now `StateTransitionGraph` is using composition only.

Reply:
Could be. The main change here is that I wanted to stabilize the interface and make the node and edge props private. Theoretically the implementation underneath could change as long as it adapts to the interface at the end